logic_mod/Blocks/Api/CpuMath.cs

```diff
@@ -114,9 +153,246 @@ public object Atan(VarCtx ctx, object[] x)
+            return (float)Math.Exp(v);
+        }
+
+        public bool WithinTol(float x, float y, float atol, float rtol)
+        {
+            return Math.Abs(x - y) <= atol + rtol * Math.Abs(y);
+        }
+
+        // Find a zero of a real or complex function using the Newton-Raphson
+        // (or secant or Halley’s) method.
+        //
+        // Math.newton(func, x0, fprime=null, tol=1.48e-08, maxiter=50, fprime2=null,
+        // x1, rtol=0.0, full_output=false)
+        //
+        // Mimicks scipy's implementation of scipy.optimize.newton
+        // 
+        // Parameters
+        // func: function
+        //      The function whose zero is wanted. It must be a function of a
+        //      single variable
+        // x0: float
+        //      An initial estimate of the zero that should be somewhere near
+        //     the actual zero.
+        // fprime : function, optional
+        //      The derivative of the function when available and convenient. If it
+        //      is null (default), then the secant method is used.
+        // tol : float, optional
+        //      The allowable error of the zero value.
+        // maxiter : int, optional
+        //      Maximum number of iterations.
+        // fprime2 : function, optional
+        //      The second order derivative of the function when available and
+        //      convenient. If it is null (default), then the normal Newton-Raphson
+        //      or the secant method is used. If it is not null, then Halley's method
+        //      is used.
+        // x1 : float, optional
+        //      Another estimate of the zero that should be somewhere near the
+        //      actual zero. Used if `fprime` is not provided.
+        // rtol : float, optional
+        //      Tolerance (relative) for termination.
+        // full_output : bool, optional
+        //      If `full_output` is false (default), the root is returned.
+        //      If true, the dictionary {{"root": root}, {"converged": true/false},
+        //      {"iter": numIter}} is returned.
+        public object Newton(VarCtx ctx, object[] x)
+        {
+            int l = x.Length;
+
+            // Arguments and their default values:
+            object func;
+            float x0;
+            object fprime = null;
+            float tol = 1.48e-08F;
+            long maxiter = 50;
+            object fprime2 = null;
+            float x1 = 0;
+            float rtol = 0.0F;
+            bool full_output = false;
+
+            bool x1Provided = false;
+
+            if (l < 2 || l > 9)
+                throw new Exception("Invalid value");
+
+            // Conditionally initialize the arguments
+            int curArgIndex = 0;
+
+            func = x[curArgIndex++];
+
+            if (!BlockUtils.TryGetFloat(x[curArgIndex++], out x0))
+                throw new Exception("Invalid value");
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
```

Comment:
Please avoid using "goto" in C#.
For example, you can combine all arguments in a struct and parse them in separate method, so you can use "return resultStruct;" instead of goto. Also you can declare a local function "void Parse() {...}" right here and use it.